### PR TITLE
docs: add aux file v2 RFC

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

aux v2 migration is near the end and I rewrote the RFC based on what I proposed (several months before...) and what I actually implemented.

## Summary of changes

Add aux file v2 RFC

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by skyzh** [comment]
I hope I've covered all key design decisions throughout the process, if not, I can add more to the RFC

File: docs/rfcs/038-aux-file-v2.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+* Image layer creation: instead of calling `key.next()` and getting/reconstructing images for every single key, we use the vectored get API to scan all keys in the keyspace at a given LSN. Image layers are only created if there are too many delta layers between the latest LSN and the last image layer we generated for sparse keyspaces.
+
+## Migration
+
+During the period of migration, we store the aux policy in the `index_part.json` file. When a tenant is attached
+with no policy set, the pageserver will scan the aux file keyspaces to identify the current aux policy being used (v1 or v2).
+
+If a timeline has aux v1 files stored, it will use aux file policy v1 unless we do a manual migration for them. Otherwise, the default aux file policy for new timelines is aux v2.
+
+The code for aux v2 migration is in https://github.com/neondatabase/aux_v2_migration. The toolkit scans all projects with logical replication enabled. For all these projects, it put the computes into maintenance mode (suspend all of then), call the migration API to switch the aux file policy on the pageserver (which drops all replication states), and restart all the computes.
```
Comment (problame) [inline docs/rfcs/038-aux-file-v2.md:108]: You need to state somewhere, perhaps in a preceding paragraph, that the only user of auxv1 are logical replication beta users.

**Review by problame — COMMENTED** [review]
Good read!

I think there's no shame in pointing out at the beginning of the doc that this is a retro-active RFC.

Also, point out in Motivation that this RFC should help understand the migration code (`aux_v2_migration.git`)

File: docs/rfcs/038-aux-file-v2.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+the deltas at each of the LSNs. However, this improved v1 storage strategy still
+requires us to store everything in an aux file cache in memory, because we cannot
+fetch a single key (or file) from the compound `AUX_FILES_KEY`.
+
+We also decided not to make the new aux storage strategy (v1) compatible with the
+original one (v1). One feasible way of doing a seamless migration is to store new
+data in aux v2 while old data in aux v1, but this complicates file deletions. We
+want all users to start with a clean state with no aux files in the storage, and
+therefore, we need to do manual migrations for users using aux v1 by using the
+[migration script](https://github.com/neondatabase/aux_v2_migration).
```
Comment (VladLazar) [inline docs/rfcs/038-aux-file-v2.md:23]: nit: I find a bit odd that this paragraph is under `Motivation`.

File: docs/rfcs/038-aux-file-v2.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+Pageserver uses fixed-size keys. The key is 128b. In order to store files of arbitrary filenames into the
+keyspace, we assign a predetermined prefix based on the directory storing the aux file, and use the FNV hash
+of the filename for the rest bits of the key. The encoding scheme is defined in `encode_aux_file_key`.
+
+For example, `pg_logical/mappings/test1` will be encoded as:
+
+```
+62 0000 01 01 7F8B83D94F7081693471ABF91C
+^ aux prefix
+        ^ assigned prefix of pg_logical/
```
Comment (VladLazar) [inline docs/rfcs/038-aux-file-v2.md:70]: Assigned by whom/what?

File: docs/rfcs/038-aux-file-v2.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+*  We use the vectored get API to retrieve all aux files during generating the basebackup. Because we need to scan a sparse keyspace, we slightly modified the vectored get path. The vectorized API will attempt to retrieve every single key within the requested key range, and therefore, we modified it in a way that keys within `NON_INHERITED_SPARSE_RANGE` will not trigger missing key error.
+
+## Compaction and Image Layer Generation
+
+With the add of sparse keyspaces, we also modified the compaction code to accommodate the fact that sparse keyspaces do not have every single key stored in the storage.
+
+* L0 compaction: we modified the hole computation code so that it can handle sparse keyspaces when computing holes.
+* Image layer creation: instead of calling `key.next()` and getting/reconstructing images for every single key, we use the vectored get API to scan all keys in the keyspace at a given LSN. Image layers are only created if there are too many delta layers between the latest LSN and the last image layer we generated for sparse keyspaces.
```
Comment (VladLazar) [inline docs/rfcs/038-aux-file-v2.md:106]: `too many delta layers`: I'm a bit confused by this (and generally image layer generation for aux files). These image layers are a subset of the aux keyspace, right? Hence, nothing should overlap with them apart from L0s which cover the whole keyspace and other AUX image layers. 

While writing this I realised I don't know if we have only one image layer for aux keys at a given LSN or multiple. Perhaps you could add a bit more detail on how the keys are represented in storage?

**Review by problame — APPROVED** [review]

File: docs/rfcs/038-aux-file-v2.md (hunk 4 above)
Comment (skyzh) [reply docs/rfcs/038-aux-file-v2.md:101]: currently we generate image layers covering the full aux keyspace, which can be optimized later :)

**Pull request merged**